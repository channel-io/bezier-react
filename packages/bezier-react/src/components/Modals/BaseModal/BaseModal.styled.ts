/* Internal dependencies */
import {
  styled,
  css,
  TransitionDuration,
  RoundAbsoluteNumber,
  keyframes,
} from 'Foundation'

const visibleAnimation = keyframes`
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
`

const translateAnimation = keyframes`
  from {
    transform: translate(0, -122px);
  }

  to {
    transform: translate(0, 0);
  }
`

export const BaseModalWrapper = styled.div<{ padded?: boolean }>`
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;

  ${({ padded }) => padded && css`
    padding: 40px 0;
  `}

  animation-name: ${() => visibleAnimation};
  animation-duration: ${TransitionDuration.M}ms;
`

export const BaseModalBackgroundOverlay = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: ${({ foundation }) => foundation?.theme?.['bgtxt-absolute-black-lighter']};
`

export const BaseModalChildrenScrollArea = styled.div`
  position: relative;
  z-index: 1;
  min-width: 360px;
  max-height: 100%;
  overflow-y: auto;
  color: ${({ foundation }) => foundation?.theme['bg-grey-darkest']};
  background-color: ${({ foundation }) => foundation?.theme?.['bg-white-high']};
  border-radius: ${RoundAbsoluteNumber.R20}px;

  animation-name: ${() => translateAnimation};
  animation-duration: ${TransitionDuration.M}ms;
  ${({ foundation }) => foundation?.elevation.ev4()};
`
