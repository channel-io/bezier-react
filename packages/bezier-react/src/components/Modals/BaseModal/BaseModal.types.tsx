import { BezierComponentProps } from 'Types/ComponentProps'

type BaseModalHtmlAttributes = React.HTMLAttributes<HTMLDivElement>

export interface BaseModalProps extends
  BaseModalHtmlAttributes,
  BezierComponentProps {
  /**
   * @description
   */
  show: boolean

  /**
   * Callback to close modal
   */
  onHide: () => void

  /**
   * Focus when this is opened
   * @default true
   */
  autoFocus?: boolean

  /**
   * When this is true, added modal padding-x
   * @default true
   */
  padded?: boolean

  /**
   * background overlay color classname
   */
  backdropClassName?: string

  /**
   * Portal's target element
   * @default getRootElement()
   */
  targetElement?: HTMLElement | null

  /**
   * @default 1e7
   */
  zIndex?: number
}
