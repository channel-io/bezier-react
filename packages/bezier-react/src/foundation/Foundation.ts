/* Internal dependencies */
import { Theme } from './Colors'
import { Spacing } from './Spacing'
import { Transition } from './Transition'
import { Rounding } from './Rounding'
import { Elevation } from './Elevation'
import { Border } from './Border'

export interface Foundation {
  theme: Theme
  subTheme?: Theme
  spacing: typeof Spacing
  transition: typeof Transition
  rounding: typeof Rounding
  elevation: typeof Elevation
  border: typeof Border
}
