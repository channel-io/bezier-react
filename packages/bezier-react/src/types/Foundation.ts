import {
  type Foundation,
  type css,
} from '~/src/foundation'

export interface FoundationProps {
  foundation?: Foundation
}

export type InjectedInterpolation = ReturnType<typeof css>

export interface InterpolationProps {
  interpolation?: InjectedInterpolation
}
