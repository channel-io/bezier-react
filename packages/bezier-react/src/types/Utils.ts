export type Entries<Type> = {
  [Key in keyof Type]: [Key, Type[Key]]
}[keyof Type][]
