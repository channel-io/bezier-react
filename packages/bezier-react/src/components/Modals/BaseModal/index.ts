import BaseModal from './BaseModal'
import type { BaseModalProps } from './BaseModal.types'

export type {
  BaseModalProps,
}
export {
  BaseModal,
}
