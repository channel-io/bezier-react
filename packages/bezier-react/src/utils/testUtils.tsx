/* External dependencies */
import React from 'react'
import {
  render as baseRender,
  renderHook as baseRenderHook,
} from '@testing-library/react'

/* Internal dependencies */
import { LightFoundation } from '~/src/foundation'
import BezierProvider from '~/src/providers/BezierProvider'
import { type ChildrenProps } from '~/src/types/ComponentProps'

function TestProviders({ children }: ChildrenProps) {
  return (
    <BezierProvider foundation={LightFoundation}>
      { children }
    </BezierProvider>
  )
}

export function render(
  ui: Parameters<typeof baseRender>[0],
  options?: Parameters<typeof baseRender>[1],
) {
  return baseRender(ui, {
    wrapper: TestProviders,
    ...options,
    legacyRoot: false,
  })
}

interface RenderHookOptions<Props> {
  initialProps?: Props
  wrapper?: React.JSXElementConstructor<{ children: React.ReactElement }>
}

export function renderHook<Result, Props>(
  hook: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props>,
) {
  return baseRenderHook(hook, {
    wrapper: TestProviders,
    ...options,
  })
}
