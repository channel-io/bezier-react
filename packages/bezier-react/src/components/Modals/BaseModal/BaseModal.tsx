/* External dependencies */
import React, {
  forwardRef,
  useCallback,
  useEffect,
  useRef,
} from 'react'
import ReactDom from 'react-dom'

/* Internal dependencies */
import { document, getRootElement } from 'Utils/domUtils'
import * as Styled from './BaseModal.styled'
import type { BaseModalProps } from './BaseModal.types'

function BaseModal(
  {
    children,
    show,
    onHide,
    autoFocus = true,
    padded = true,
    backdropClassName = '',
    targetElement = getRootElement(),
    zIndex = 1e7,
    style,
    ...rests
  }: BaseModalProps,
  forwardedRef: React.Ref<HTMLDivElement>,
) {
  const baseModalBodyRef = useRef<HTMLDivElement>(null)

  const onClickBackgroundOverlay = useCallback((e: React.MouseEvent<HTMLDivElement>) => {
    e.stopPropagation()
    onHide()
  }, [onHide])

  const onCloseModalByEsc = useCallback((e: KeyboardEvent) => {
    switch (e.key) {
      case 'Escape': {
        e.stopPropagation()
        onHide()
        break
      }
    }
  }, [onHide])

  useEffect(function initCloseEvent() {
    if (show) {
      document.addEventListener('keydown', onCloseModalByEsc)
      return function cleanup() {
        document.removeEventListener('keydown', onCloseModalByEsc)
      }
    }
    return undefined
  }, [
    show,
    onCloseModalByEsc,
  ])

  useEffect(function initAutoFocus() {
    if (autoFocus && show && baseModalBodyRef.current) {
      baseModalBodyRef.current.focus()
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [show])

  if (!targetElement || !show) { return null }

  return (
    ReactDom.createPortal(
      <Styled.BaseModalWrapper
        ref={forwardedRef}
        data-testid="BaseModal"
        padded={padded}
        style={{
          zIndex,
        }}
      >
        <Styled.BaseModalBackgroundOverlay
          className={backdropClassName}
          data-testid="BaseModal__Background"
          onClick={onClickBackgroundOverlay}
        />
        <Styled.BaseModalChildrenScrollArea
          tabIndex={-1}
          data-testid="BaseModal__Children"
          {...rests}
          ref={baseModalBodyRef}
        >
          { children }
        </Styled.BaseModalChildrenScrollArea>
      </Styled.BaseModalWrapper>,
      targetElement,
    )
  )
}

export default forwardRef(BaseModal)
