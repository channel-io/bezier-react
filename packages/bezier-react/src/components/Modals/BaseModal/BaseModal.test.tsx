/* External dependencies */
import React from 'react'
import type { RenderResult } from '@testing-library/react'
import { fireEvent } from '@testing-library/react'

/* Internal dependencies */
import { LightFoundation } from 'Foundation'
import { render } from 'Utils/testUtils'
import BaseModal from './BaseModal'
import type { BaseModalProps } from './BaseModal.types'

const DEFAULT_PROPS: BaseModalProps = {
  show: true,
  className: 'BaseModal',
  onHide: jest.fn(),
}

describe('BaseModal', () => {
  let renderResult: RenderResult

  describe('When "show" is false', () => {
    beforeEach(() => {
      renderResult = render(
        <BaseModal {...DEFAULT_PROPS} show={false} />,
        {
          container: document.body,
        },
      )
    })

    afterEach(jest.clearAllMocks)

    it('Component is "null"', async () => {
      const target = renderResult.container.firstChild
      expect(target).toBe(null)
    })
  })

  describe('When "show" is true', () => {
    beforeEach(() => {
      renderResult = render(
        <BaseModal {...DEFAULT_PROPS} />,
        {
          container: document.body,
        },
      )
    })

    afterEach(jest.clearAllMocks)

    it('should equal the snapshot', async () => {
      const target = await renderResult.findByTestId('BaseModal')
      expect(target).toMatchSnapshot()
    })

    it('BaseModal is mounted', async () => {
      const target = await renderResult.findByTestId('BaseModal')
      expect(target).toBeInTheDocument()
    })

    it('className is transferred into BaseModal__Children', async () => {
      const target = await renderResult.findByTestId('BaseModal__Children')
      expect(target.classList.contains(DEFAULT_PROPS.className || '')).toBe(true)
    })

    it('When Modal__Background is clicked, the onHide is called', async () => {
      const target = await renderResult.findByTestId('BaseModal__Background')
      fireEvent.click(target)
      expect(DEFAULT_PROPS.onHide).toBeCalledTimes(1)
    })

    it('When "Escape" key is down, the onHide is called', async () => {
      const target = await renderResult.findByTestId('BaseModal__Background')
      fireEvent.keyDown(target, { key: 'Escape' })
      expect(DEFAULT_PROPS.onHide).toBeCalledTimes(1)
    })
  })

  describe('When "autoFocus" is true', () => {
    beforeEach(() => {
      renderResult = render(
        <BaseModal
          {...DEFAULT_PROPS}
          autoFocus
        />,
        {
          container: document.body,
        },
      )
    })

    afterEach(jest.clearAllMocks)

    it('BaseModal__Children has focus', async () => {
      const target = await renderResult.findByTestId('BaseModal__Children')
      expect(target).toHaveFocus()
    })
  })

  describe('Styles', () => {
    beforeEach(() => {
      renderResult = render(
        <BaseModal {...DEFAULT_PROPS} />,
        {
          container: document.body,
        },
      )
    })

    afterEach(jest.clearAllMocks)

    it('BaseModal', async () => {
      const target = await renderResult.findByTestId('BaseModal')
      expect(target).toHaveStyle('position: fixed;')
      expect(target).toHaveStyle('top: 0;')
      expect(target).toHaveStyle('right: 0;')
      expect(target).toHaveStyle('bottom: 0;')
      expect(target).toHaveStyle('left: 0;')
      expect(target).toHaveStyle('display: flex;')
      expect(target).toHaveStyle('align-items: center;')
      expect(target).toHaveStyle('justify-content: center;')
    })

    it('BaseModal__Background', async () => {
      const target = await renderResult.findByTestId('BaseModal__Background')
      expect(target).toHaveStyle('position: absolute;')
      expect(target).toHaveStyle('top: 0;')
      expect(target).toHaveStyle('right: 0;')
      expect(target).toHaveStyle('bottom: 0;')
      expect(target).toHaveStyle('left: 0;')
      expect(target).toHaveStyle(`background-color: ${LightFoundation.theme['bgtxt-absolute-black-lighter']}`)
    })

    it('BaseModal__Children', async () => {
      const target = await renderResult.findByTestId('BaseModal__Children')
      expect(target).toHaveStyle('display: block;')
      expect(target).toHaveStyle('min-width: 360px;')
      expect(target).toHaveStyle('color: rgb(36, 36, 40);')
      expect(target).toHaveStyle('background-color: #fff;')
      expect(target).toHaveStyle('border-radius: 20px;')
      expect(target).toHaveStyle('box-shadow: inset 0 0 2px 0 #FFFFFF1F, 0 0 2px 1px #0000000D, 0 4px 20px #00000038;')
    })
  })
})
