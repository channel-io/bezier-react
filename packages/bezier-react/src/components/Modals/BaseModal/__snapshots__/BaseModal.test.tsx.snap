// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseModal When "show" is true should equal the snapshot 1`] = `
.c0 {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 40px 0;
  -webkit-animation-name: jgQpwH;
  animation-name: jgQpwH;
  -webkit-animation-duration: 300ms;
  animation-duration: 300ms;
  -webkit-animation-timing-function: cubic-bezier(0.3,0,0,1);
  animation-timing-function: cubic-bezier(0.3,0,0,1);
}

.c1 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: #00000066;
}

.c2 {
  position: relative;
  z-index: 1;
  min-width: 360px;
  max-height: 100%;
  overflow-y: auto;
  color: #242428;
  background-color: #FFFFFF;
  border-radius: 20px;
  -webkit-animation-name: cYIBgc;
  animation-name: cYIBgc;
  -webkit-animation-duration: 300ms;
  animation-duration: 300ms;
  -webkit-animation-timing-function: cubic-bezier(0.3,0,0,1);
  animation-timing-function: cubic-bezier(0.3,0,0,1);
  background-color: #FFFFFF;
  box-shadow: inset 0 0 2px 0 #FFFFFF1F, 0 0 2px 1px #0000000D, 0 4px 20px #00000038;
}

<div
  class="c0"
  data-testid="BaseModal"
  style="z-index: 10000000;"
>
  <div
    class="c1"
    data-testid="BaseModal__Background"
  />
  <div
    class="c2 BaseModal"
    data-testid="BaseModal__Children"
    tabindex="-1"
  />
</div>
`;
